### ISSUE TITLE
Optimize state diff calculation / reporting

### ISSUE THREAD

**Issue by 20k-ultra** [balena-os/balena-supervisor#1716]
The [current state](https://github.com/balena-os/balena-supervisor/blob/master/src/device-state/current-state.ts) module checks for the current state on a interval or device change which it then compares before deciding to report. I have seen that the state diffing can cause cpu spikes of 18% on my rpi4. This calculation is not throttled and can be called infinite amount of times but on a functioning device it is every 10 seconds or device change as I just mentioned.

We should optimize this function to either throttle calls or cache responses or use a different algorithm for diffing the states.

**Comment by 20k-ultra** [comment]
I have actually seen that making the diff calculation return early shows that the cpu spikes are from the gathering of the current state.

https://github.com/balena-os/balena-supervisor/blob/a4be293ec48755f7c3c75a0db1db6bacb3f628d8/src/device-state/current-state.ts#L178 causes generally 2% CPU spike..I saw it go up to 20% once but that was an anomaly.

Just below it, the system info check is responsible for consistently having 18% CPU spikes when executed. This is where we should focus our optimizations as that seems to be pretty ridiculous.

**Comment by 20k-ultra** [comment]
So the getSystemMetrics function has been seen to spike the CPU to 18% usage when executed. Here are CPU usage trends when only the individual metrics are calculated:

 - cpuUsage: does not register any significant cpu usage (< 1%)
 - cpuTemp: consistently raised cpu usage by 6-8%
 - getMemoryInformation shows minimal cpu usage (1% at most)
 - getCpuId: does not register any significant cpu usage (< 1%)
 - getStorageInfo: consistently raised cpu usage by 4%

**Comment by cywang117** [comment]
Regarding metrics reporting, since `cpuTemp` and `getStorageInfo` are the main perpetrators, maybe we can find more linux-native solutions? I took a look at `getStorageInfo` and it seems to just be reporting the storage used, storage total, and storage block device of the `/mnt/data` mount:
```
$ df -h
Filesystem                               Size  Used Avail Use% Mounted on
/dev/nvme0n1p6                  227G  2.0G  214G   1% /mnt/data
/dev/nvme0n1p1                   40M  3.8M   37M  10% /mnt/boot
/dev/nvme0n1p3                  300M  2.1M  278M   1% /mnt/sysroot/inactive
```
![image](https://user-images.githubusercontent.com/38748558/120144761-7eac6700-c21d-11eb-8c8f-1db6a37a3e07.png)
This is also evidenced in the current [code](https://github.com/balena-os/balena-supervisor/blob/dcd863eed8b58cc7eae866c941597c7e0a9bfcb1/src/lib/system-info.ts#L25) itself.

It's possible that we could have less spikes for `getStorageInfo` by switching to `child_process.exec("df -h | grep '/mnt/data'")` and parsing those results.

As for CPUTemp, [here](https://github.com/sebhildebrandt/systeminformation/blob/master/lib/cpu.js#L931) is `systeminformation`'s implementation. I'm wondering if it's possible, or preferable even, to extract something that's slimmed down from that implementation. A direct linux command might not work because differences arch types storage temperature info in different places (i.e. `/sys/devices/platform/coretemp.0/hwmon/hwmon*/temp*_input` instead of `/sys/class/thermal/thermal_zone*/temp` for Intel NUC)

**Comment by 20k-ultra** [comment]
It doesn't seem like node offers a standard lib approach for storage devices so executing df -h seems like the most reasonable.

For the cpuTemp we could use some of our own heuristics and say if we know the top 5 DTs that implement CPU temp differently then we can support those and default to using systeminformation if we cannot get it. In the happy case we get better performance with the robustness of the more developed module.

**Comment by pipex** [comment]
@20k-ultra Is this issue still relevant? I think with all the improvements we have introduced there is no much to do here right?

**Comment by 20k-ultra** [comment]
There are 2 points about this; current state calculation might be called more times than necessary, and, the calculation cost of the current state difference from the last reported state might be optimized.

I think https://github.com/balena-os/balena-supervisor/commit/55a344dceb826c5655067e6006e006846778b5e9 is the only change that prevented excessive CPU usage. It improved the situation because before the Supervisor would calculate state diff, send diff (I think?), then check what the state diff was again. If the CPU temp changed or mem usage went up a bit within that sequence, it (Supervisor) would check state diff again and get stuck in this loop. Now the SV just sends the report and waits until the next report tick comes around instead of checking what was just reported is what the device currently has.

The current reporting might still try to calculate the current state in quick succession since the only thing throttling this calculation is if a report was sent. One scenario I can think of is, if the current state is the same as last reported state the report throttle is not set.

Aside from that, there also haven't been any improvements to the functions that calculate the state diff. I mentioned in https://github.com/balena-os/balena-supervisor/issues/1716#issuecomment-850728254 which parts cost the most.

Realistically, there are no changes needed but I wanted to document the potential for improvement.

**Comment by pipex** [comment]
Right, I thought we had throttled the "query" part of the report, but we improved the target state report instead. I think what is confusing is that the issue description is not super clear, the description mentions the specific bug, which has been fixed, but it also points to the diffing which won't make much sense to anyone that has not worked on that specific module.

**Comment by kb2ma** [comment]
#2049 reduces CPU time by throttling metrics-only measurement/reporting to once per 5 minutes. See [this comment](https://github.com/balena-os/balena-supervisor/issues/2007#issuecomment-1325851246) for more discussion. So that aspect of this issue has been addressed.
